File: ruby/algorithms/exp3/exp3.rb
class Exp3
  def initialize(gamma, n_arms)
    @gamma = gamma
    @weights = Array.new(n_arms, 1.0)
  end

  def reset(n_arms)
    @weights = Array.new(n_arms, 1.0)
  end

  def select_arm
    n_arms = @weights.size()
    total_weight = @weights.reduce(:+)
    probs = Array.new(n_arms, 0.0)
    n_arms.times do |arm|
      probs[arm] = (1 - @gamma) * (@weights[arm] / total_weight)
      probs[arm] = probs[arm] + @gamma * (1.0 / n_arms.to_f)
    end
    categorical_draw(probs)
  end

  def update(chosen_arm, reward)
    n_arms = @weights.size()
    total_weight = @weights.reduce(:+)
    probs = Array.new(n_arms, 0.0)
    n_arms.times do |arm|
      probs[arm] = (1 - @gamma) * (@weights[arm] / total_weight)
      probs[arm] = probs[arm] + @gamma * (1.0 / n_arms.to_f)
    end

    x = reward / probs[chosen_arm]

    growth_factor = Math.exp((@gamma / n_arms) * x)
    @weights[chosen_arm] = @weights[chosen_arm] * growth_factor
  end

  private
  def categorical_draw(probs)
    z = rand()
    cum_prob = 0.0

    probs.size().times do |i|
      prob = probs[i]
      cum_prob += prob
      if cum_prob > z
        return i
      end
    end

    return probs.size() - 1
  end
end

File: julia/algorithms/exp3/exp3.jl
type Exp3 <: BanditAlgorithm
  gamma::Float64
  weights::Vector{Float64}
end

function Exp3(n_arms::Int64, gamma::Float64)
  Exp3(gamma, ones(n_arms))
end

function initialize(algo::Exp3, n_arms::Int64)
  algo.weights = ones(n_arms)
end

function initialize(algo::Exp3, arms::Vector)
  initialize(algo, length(arms))
end

function select_arm(algo::Exp3)
  K = length(algo.weights)
  probs = (1 - algo.gamma) * (algo.weights ./ sum(algo.weights)) + (algo.gamma / K)
  return rand(Categorical(probs))
end

function update(algo::Exp3, chosen_arm::Int64, reward::Real)
  K = length(algo.weights)
  probs = (1 - algo.gamma) * (algo.weights ./ sum(algo.weights)) + (algo.gamma / K)

  x = zeros(K)
  x[chosen_arm] = reward / probs[chosen_arm]

  algo.weights = algo.weights .* exp((algo.gamma / K) .* x)  
end
